File: proyecto/cppcalc/parser.cpp
#include "parser.h"
#include "calcex.h"
#include "calculator.h"
#include <string>
#include <sstream>

string resultExtl = "";

Parser::Parser(istream* in) {
   scan = new Scanner(in);
}

Parser::~Parser() {
   try {
      delete scan;
   } catch (...) {}
}

AST* Parser::parse() {
   return Prog();
}

AST* Parser::Prog() {
   AST* result = Expr();
   Token* t = scan->getToken();

   if (t->getType() != eof) {
     cout << "Syntax Error: Expected EOF,found token at column " <<
       t->getCol() << endl;
     throw ParseError;
   }
   
   return result;
}

AST* Parser::Expr() {
  return RestExpr(Term());
}

AST* Parser::RestExpr(AST* e) {
  
  
  Token* t = scan->getToken();
  
  if (t->getType() == add) {
    return RestExpr(new AddNode(e,Term()));
   }
  
  if (t->getType() == sub)
    return RestExpr(new SubNode(e,Term()));
  
  if(t->getType() == mod)
    return RestExpr(new ModNode(e,Term()));
  
  scan->putBackToken();

  return e;
}

AST* Parser::Term() {
  
  return RestTerm(Storable());
  
}

AST* Parser::RestTerm(AST* e) {
  Token* t = scan->getToken();
  
  if (t->getType() == times) {
    return RestTerm(new TimesNode(e,Storable()));
  }
  
  if (t->getType() == divide)
    return RestTerm(new DivideNode(e,Storable()));
  
  scan->putBackToken();
  
  return e;
  
  
  throw ParseError; 
}

AST* Parser::MemOperation() {
  AST *result = Factor();
  
  Token *t = scan -> getToken();
  
  if(t -> getType() == keyword)
    if(t -> getLex() == "S")
      return new StoreNode(result);
    else if(t -> getLex() == "P"){
      return new PlusNode(result);
    }else if(t -> getLex() == "M"){
      return new MinusNode(result);
    }else{
      cout << "Syntax Error: found other keyword at pos " 
	   << t -> getCol()
	   << endl;
      throw ParseError;
    }
  scan -> putBackToken();
  return result;
  
  
}

AST* Parser::Storable(){
  return MemOperation();
} 

AST* Parser::Factor() {
  Token* t = scan->getToken();
  
  if (t->getType() == number) {
    istringstream in(t->getLex());
    int val;
    in >> val;
    return new NumNode(val);
  }else if(t -> getType() == identifier){
    istringstream in(t->getLex());
    string val;
    in >> val;
    return (Assignable(val));
  }
    
  
  if(t -> getType() == keyword){
    if(t -> getLex() == "R"){
      return new RecallNode();
    }else if( t -> getLex() == "C"){
      return new ClearNode();
    }
    else{
      cout << "Syntax error Keyword"
	   << t -> getLex()
	   << "at line " << t -> getLine()
	   << "col : " << t -> getCol() << endl;
      throw ParseError;
    }
  }
  
  if(t -> getType() == lparen){
    AST *result = Expr();
    t = scan -> getToken();
    if( t -> getType() == rparen){
      return result;
    }
    cout << "Syntax error expected ) at line:  " <<
      t -> getLine()<< "col : " <<
      t -> getCol() << endl;
    throw ParseError;
  }
  
  cout << "Syntax error expected number, R , ( at line " <<
    t -> getLine()<< "col : " <<
    t -> getCol() << endl;
  
  throw ParseError;
}

AST* Parser::Assignable(string n){
  // cout << "me llego a asignable"<< endl;  
  return Assign(n);
}

AST* Parser::Assign(string n){
  //cout << "me llego a asign"<< endl;
  Token *t = scan -> getToken();
  NumNode* result = new NumNode(0);
  stringstream stream;
  if(t -> getType() == equals){
    scan -> putBackToken();
    t = scan -> getToken();
    result = new NumNode(Expr() -> evaluate());
    result -> assignate(n,result -> evaluate());
    stream << result -> evaluate();
    resultExtl += n + " <- " + "," +  stream.str();
    resultExtl += result -> evaluate();
  }else{
    // cout << "me entro al esle de assign" << endl;
    int prueba = result -> search(n);
    // cout <<"else prueba " <<  prueba << endl;
    result  =  new NumNode(prueba);
    result -> assignate(n,prueba);
  }
  scan -> putBackToken();
  return result;
}

File: proyecto/rubycalc/lib/parser.rb
require 'ast'
require 'scanner'
require 'token'
require 'calcex'

class Parser
  def initialize(istream)
    @scan = Scanner.new(istream)
  end
   
  def parse()
    return Prog()
  end
  
  private
  def Prog()
    result = Expr()
    t = @scan.getToken()
    
    if t.type != :eof then
      print "Expected EOF. Found ", t.type, ".\n"
      raise ParseError.new
    end
  
     result
  end
  
  def Expr() 
    RestExpr(Term())
  end
  
  def RestExpr(e) 
    t = @scan.getToken
    
    if t.type == :add then
      return RestExpr(AddNode.new(e,Term()))
    end
    
    if t.type == :sub then
      return RestExpr(SubNode.new(e,Term()))
    end
    
    @scan.putBackToken    
     e

  end
  
  def Term
    RestTerm(Storable())  
  end
   
  def RestTerm(e)
    t = @scan.getToken

    if t.type == :times then
      return RestTerm(TimesNode.new(e,Storable()))
    end

    if t.type == :divide then
      return RestTerm(DivideNode.new(e,Storable()))
    end
    if t.type == :mod then
        return RestTerm(ModNode.new(e,Storable()))
    end    
    @scan.putBackToken
    return e
   
  end
   
  def Storable()
    result = Factor()
    return MemOperation(result)
  end

  def MemOperation(result)
    t = @scan.getToken
    if t.type == :keyword then
      if t.lex == "S" 
        return StoreNode.new(result)
      elsif t.lex == "M"
        return MinusNode.new(result)
      elsif t.lex == "P"
        return PlusNode.new(result)
      end

      puts "Expected S or M or P  found #{t.type} at line #{t.line}"
      raise ParseError.new   
    end
    @scan.putBackToken
    result
  end
   
  def Assignable
    t = @scan.getToken
    Assign(t.lex.to_s)
  end
  
  def Assign lex
    t = @scan.getToken
    #result = NumNode.new(0)
    if t.type == :equals
      result = NumNode.new(Expr().evaluate())
      result.assign lex
    else
      @scan.putBackToken
      result = NumNode.new(0)
      result.search lex    
    end
    return result
  end

  def Factor() 
    t = @scan.getToken
    
    if t.type == :number then
      return NumNode.new t.lex.to_i
    end
    
    if t.type == :identifier then
      @scan.putBackToken
      return Assignable()
    end
    
    if t.type == :keyword then
      if t.lex == "R" 
        return RecallNode.new
      elsif t.lex == "C"
        return ClearNode.new
      end
      puts "Parse Error: expected R found: " + t.lex
      puts "at line: " + t.line.to_s + "col: " + t.col.to_s
      raise ParseError.new
    end
    
    if t.type == :lparen then
      result = Expr()
      t = @scan.getToken
      if t.type == :rparen then
        return result
      end
      puts "Parse Error "
      puts "at line: " + t.line.to_s + "col: " + t.col.to_s
      raise ParseError.new
    end
  
    print "Parse "
    puts "at line: " + t.line.to_s + "col: " + t.col.to_s
    raise ParseError.new
  end
end
